### ISSUE TITLE
[ENH] `SARIMAX` - implement `predict_interval` or `predict_quantiles` (for `statsmodels` `SARIMAX` interface)

### ISSUE THREAD

**Issue by fkiraly** [sktime/sktime#4301]
From discord, Huseyin Aytug:

hi when I want to produce confidence intervals with `SARIMAX`, I get this error:

`NotImplementedError: SARIMAX does not have the capability to return quantile predictions. If you think this estimator should have the capability, please open an issue on sktime.`

However, it should be possible according to the documentation http://www.sktime.net/en/latest/api_reference/auto_generated/sktime.forecasting.sarimax.SARIMAX.html#sktime.forecasting.sarimax.SARIMAX.predict_interval

**Comment by fkiraly** [comment]
I think Huseyin is entirely correct - this should be possible in-princinple.

Other classes with ARIMA models inside could be used as a template, e.g., the `pmdarima` based `ARIMA` and `AutoARIMA`, or the `statsforecast` based `StatsforecastAutoARIMA`, see the `_predict_interval` or `_predict_quantiles` methods there.

I think @kejsitake implemented most of these, if I remember correctly.

**Comment by yarnabrina** [comment]
Hi, I am interested in working on this functionality, if no one else is working.

**Comment by fkiraly** [comment]
👍 I think no one is working on it currently.

**Comment by fkiraly** [comment]
There's also quite a lot of copy-paste redundancy with minor variations going on, plus these two bugs:

* https://github.com/sktime/sktime/issues/4363
* https://github.com/sktime/sktime/issues/4362

(all in `statsmodels` interfaces)

so if you have any good ideas about reducing multiplication/redundancy or the bugs - perhaps this can be fixed in one swoop if one has the right idea? - would be much appreciated. Not necessary of course, just mentioning what is going on in the area.

**Comment by yarnabrina** [comment]
I was going through `pmdarima` codebase, and I see they have a separate handling in case user chooses to use `simple_differencing`.

https://github.com/alkaline-ml/pmdarima/blob/18cbe34d62bf7ec2508509b0590ff0962f00af3e/pmdarima/arima/arima.py#LL168-L171

I could not find such handling in `_predict` for `SARIMAX` in `sktime`. Any idea if it is an unintended omission, or if it is being handled somewhere else?

FWIW, I tried running `test_SARIMAX_against_statsmodels` adding this option, but at least for the generated `df`, the predictions remain unchanged (between `True` and `False`, not just between `sktime` wrapper and underlying implementation). So, now I am wondering whether to add this extra handling for current main task (`predict_interval` for `SARIMAX` wrapper in `sktime`) or not.

**Comment by yarnabrina** [comment]
@fkiraly I am planning to work on the refactoring of `_predict_interval` over the next week. This is my idea so far before starting:

1. add a `_predict_interval` to `_StatsModelsAdapter`
    1. similar to the common pattern in `ets`, `sarimax` and `structural`
    2. one extra abstract/non-implemented method to get quantile information from corresponding `PredictionResults` class, which seem to vary between above three
3. override that in subclasses:
    1. no override in classes for which `capability:pred_int` is `False`
    2. for estimators other than above the above three, override `_predict_interval` with current methods
    3. for above three estimators, implement the new method

Please let me know if it seems okay, or if you have some other suggestions.

I have few other doubts as well:

1. What can I expect at to be commit order to start with? Shall I start with `main`, or with #4439, or with #4427, or with both of these?
2. Is it safe to assume classes inheriting `_StatsModelsAdapter` are all the estimators in the scope of this refactoring? Or, can there be some others as well?
3. Are #4362 and #4363 related to this refactoring? Or they are indexing bugs in `_predict_interval`, but not affecting this?

**Comment by fkiraly** [comment]
NICE!

I've moved the discussion to a new issue
https://github.com/sktime/sktime/issues/4447
since this one is about `SARIMAX`, and there will probably be quite some specific discussion to the general inheritance structure.
